### ISSUE TITLE
Maintenance job getting ~2000 times the same file, with huge performance and cost impact

### ISSUE THREAD

**Issue by rducom** [kopia/kopia#3059]
Hello

While running a full maintenance job, we're having the following `GetBlob` call **1937** times : 

`2023-06-08T07:58:40.918587Z DEBUG kopia/repo [STORAGE] GetBlob	{"blobID":"q894b24492b1c49d925c43a851a35fc3e-s432977430d515f6311b","offset":0,"length":-1,"outputLength":21143278,"error":null,"duration":"2.0008945s"}`

The file is ~20Mb.
First log : `2023-06-08T07:12:34.026410Z`
Last log : `2023-06-08T07:58:40.918587Z`

So Kopia is spending **46 minutes** getting the same 20MB files 1937 times, consuming **38Gb** of bandwidth on our S3.
We are seeing this behavior for all the blob files.

The last successful maintenance job on this bucket have succed in more than 72 hours.

We are using Scaleway S3 Object Storage.

Is there a way to have a local cache for the maintenance job, and avoid having so many wasted time and bandwidth ?

**Comment by jkowalski** [comment]
That is not expected and the file should be cached after first download. Are you seeing any cache errors?

**Comment by rducom** [comment]
Hi @jkowalski 
No we're only seeing `DEBUG kopia/repo [STORAGE] GetBlob` lines in the logs

**Comment by rducom** [comment]
We have tried with a dedicated KOPIA_CACHE_DIRECTORY, but this doesn't improve the behavior.

**Comment by arouene** [comment]
I work with @rducom, and found that the "issue" is when the metadata cache became full, resulting in redundant downloads of Blobs during full maintenance mode while searching for active contents. The cache was unable to accommodate these Blobs, leading to inefficient resource utilization.

What perplexed us was the absence of any log messages regarding the full cache, even in debug level mode. To address this issue, I have submitted a PR. Can you please review the PR #3085 ?

**Comment by nickdford** [comment]
We have just experienced this issue across a large number of servers after updating Kopia. Also resulting an a huge cost impact.
The symptom we spotted was that the network interface on the server maxes out and the maintenance never completes.
Is there any way to prevent this from happening?

Thanks!

**Comment by arouene** [comment]
@nickdford did you checked that your cache isn't full? What you describe is similar to our own issue, and was solved by resizing the cache.

You can check the cache usage with `kopia cache info`

**Comment by nickdford** [comment]
@arouene correct, increasing the cache limits is our workaround for the issue.
We just need a fix for the bug, as this is happening on random servers as their caches fill up. By the time we detect it, the large S3 costs will have already been racked up for the hosts.

I understand that if the cache is not large enough it will download files from the repository, but it should not be endless looping downloading the same file.

**Comment by arouene** [comment]
Well, I don´t think this is a bug. If the cache is too small, it will needs to download the same blobs each time it needs them, as it cannot store them in the cache. 

May be the LRU might be improved.

But for now you really need to allocate a big cache, our is 300GB (260GB for metadata, 40GB for data), for 300 snapshots of about 10GB each.

**Comment by nickdford** [comment]
Thanks for your reply @arouene and clarification of how a cache works.

Surely unless the maintenance command is looping in some way, it will finish running at some point?
In my mind, expected behaviour when the cache is full, would be for it to download a few files from the repository (some maybe more than once) until it has completed the maintenance, but not the same file infinite times.

let me know if you think infinite times is expected.

**Comment by arouene** [comment]
> let me know if you think infinite times is expected.

Not infinite time, but way longer, and way more bandwidth that your are expecting. For instance, since we fixed our cache size, maintenance is taking about 2h max, but took nearly 3 or 4 days with an insufficient cache...

**Comment by julio-lopez** [comment]
Actually, this ticket should not have been closed. It was incorrectly auto-closed by a change that was merged.

**Comment by kakaroto** [comment]
I agree that it's not infinite times (we're encountering the same issue, reported in #3184) but it's a lot more than it should be in my opinion. Why would it even need to download/process that same blob 1937 times, and if it did need to pass through it multiple times, wouldn't it be "new" and therefore not getting pruned by the cache when it does a flush? Unless the entire content in the cache was traversed and pruned repeatedly 1937 times during the maintenance, which also doesn't sound right.
My guess is that there's a bug in how the cached objects are handled. I don't know how maintenance works so I don't know if it's normal behavior for it to read the contents of that same blob 1937 times during the maintenance run.

In our case, we have a 2TB bucket and it would download 20TB+ (up to 29TB sometimes) during the maintenance run. Even if it was all cached data, it seems to indicate that the objects might be read more frequently than necessary (i.e: potentially lots of unnecessary I/O on disk if the content had been cached). It might be worth considering.
Thanks for looking into this.
